File: src/adeu/markup.py
import re
from typing import List, Optional, Tuple

import structlog

from adeu.diff import trim_common_context
from adeu.models import ModifyText

logger = structlog.get_logger(__name__)


def _should_strip_markers(text: str, marker: str) -> bool:
    """
    Determines if outer markers should be stripped from text.
    Only strip if:
    1. Text starts AND ends with the marker (balanced)
    2. The inner content is "prose-like" (not a code identifier or pure symbols)
    3. There are no additional marker pairs inside (e.g., "**A** and **B**" should NOT strip)
    """
    import re

    if not text.startswith(marker) or not text.endswith(marker):
        return False

    if len(text) < len(marker) * 2:
        return False

    inner = text[len(marker) : -len(marker)]

    if not inner:
        return False

    # Check for additional markers inside - if present, don't strip
    # e.g., "**A** and **B**" has "A** and **B" inside, which contains **
    if marker in inner:
        return False

    # Inner content must have actual letter characters (not just digits/underscores/symbols)
    # This prevents stripping things like "___", "__0__"
    if not re.search(r"[a-zA-Z]", inner):
        return False

    # For double-underscore (__), be conservative:
    # Don't strip if inner looks like a code identifier (e.g., "init", "name", "main")
    # Code identifiers: only word chars, no spaces
    if marker == "__":
        # If inner has no spaces and is only word characters, it's likely code like __init__
        if re.fullmatch(r"\w+", inner):
            return False

    # For single underscore (_), only skip if it looks like snake_case (contains inner underscore)
    # _emphasis_ -> strip (prose)
    # _some_var_ -> don't strip (code identifier)
    if marker == "_":
        # If inner contains underscore, likely snake_case code identifier
        if "_" in inner:
            return False
        # If inner is a single word with no letters (like just digits), don't strip
        if re.fullmatch(r"[0-9_]+", inner):
            return False

    return True


def _strip_balanced_markers(text: str) -> tuple[str, str, str]:
    """
    Strips balanced outer formatting markers from text.
    Returns (prefix_markup, clean_text, suffix_markup).

    Only strips if the markers are truly formatting (content has word chars).
    """
    prefix_markup = ""
    suffix_markup = ""
    clean_text = text

    # Check markers in order of length (longer first to avoid ** vs * conflicts)
    markers = ["**", "__", "_", "*"]

    for marker in markers:
        if _should_strip_markers(clean_text, marker):
            prefix_markup += marker
            suffix_markup = marker + suffix_markup
            clean_text = clean_text[len(marker) : -len(marker)]
            # Only strip one level of markers
            break

    return prefix_markup, clean_text, suffix_markup


def _replace_smart_quotes(text: str) -> str:
    """Normalizes smart quotes to ASCII equivalents."""
    return text.replace("“", '"').replace("”", '"').replace("‘", "'").replace("’", "'")


def _strip_markdown_for_matching(text: str) -> Tuple[str, List[int]]:
    """
    Strips markdown formatting markers and builds a position map.
    Returns (stripped_text, position_map) where position_map[i] = original index.
    """
    result = []
    position_map = []
    i = 0

    while i < len(text):
        # Skip ** or __
        if i < len(text) - 1 and text[i : i + 2] in ("**", "__"):
            i += 2
            continue
        # Skip single * or _ that look like markdown (at word boundaries)
        if text[i] in ("*", "_"):
            prev_char = text[i - 1] if i > 0 else " "
            next_char = text[i + 1] if i < len(text) - 1 else " "
            # If at boundary (space or start/end), likely markdown
            if prev_char in (" ", "\n", "\t") or next_char in (" ", "\n", "\t"):
                i += 1
                continue

        position_map.append(i)
        result.append(text[i])
        i += 1

    return "".join(result), position_map


def _find_safe_boundaries(text: str, start: int, end: int) -> Tuple[int, int]:
    """
    Adjusts match boundaries to avoid splitting markdown formatting tokens.
    Ensures that if we consume an opening marker, we also consume the closing one,
    keeping the replacement balanced.
    """
    new_start = start
    new_end = end

    def expand_if_unbalanced(marker: str):
        nonlocal new_start, new_end

        # Get current match content
        current_match = text[new_start:new_end]

        # Check if unbalanced (odd number of markers)
        if current_match.count(marker) % 2 != 0:
            # Look in suffix first (most common case for regex consuming opening tag)
            suffix = text[new_end:]
            if suffix.startswith(marker):
                new_end += len(marker)
                return  # Re-evaluate? For now assuming simple adjacency

            # Look in prefix
            prefix = text[:new_start]
            if prefix.endswith(marker):
                new_start -= len(marker)
                return

    # Iteratively check markers.
    for _ in range(2):
        expand_if_unbalanced("**")
        expand_if_unbalanced("__")
        expand_if_unbalanced("_")
        expand_if_unbalanced("*")

    return new_start, new_end


def _refine_match_boundaries(text: str, start: int, end: int) -> Tuple[int, int]:
    """
    Refines fuzzy match boundaries to avoid greedy consumption of unbalanced markers.
    Example: "**Header.**Body" -> Regex matches "**Body".
    This function trims the leading "**" because "Body" is balanced (0 markers)
    while "**Body" is unbalanced (1 marker).
    """
    # Markers to check. Order matters (check compound markers first).
    markers = ["**", "__", "*", "_"]

    current_text = text[start:end]
    best_start, best_end = start, end

    # 1. Check Leading Noise
    for marker in markers:
        if current_text.startswith(marker):
            # Calculate balance scores
            # "Unbalanced-ness" = count % 2. 0 is perfect. 1 is bad.
            current_score = current_text.count(marker) % 2

            trimmed_text = current_text[len(marker) :]
            trimmed_score = trimmed_text.count(marker) % 2

            # If we are currently unbalanced (1) and trimming makes it balanced (0), do it.
            if current_score == 1 and trimmed_score == 0:
                best_start += len(marker)
                current_text = trimmed_text  # Update for next iteration

    # 2. Check Trailing Noise (Logic is symmetric)
    for marker in markers:
        if current_text.endswith(marker):
            current_score = current_text.count(marker) % 2
            trimmed_text = current_text[: -len(marker)]
            trimmed_score = trimmed_text.count(marker) % 2

            if current_score == 1 and trimmed_score == 0:
                best_end -= len(marker)
                current_text = trimmed_text

    return best_start, best_end


def _make_fuzzy_regex(target_text: str) -> str:
    """
    Constructs a regex pattern from target text that permits:
    - Variable whitespace (\\s+)
    - Variable underscores (_+)
    - Smart quote variation
    - Intervening markdown formatting (**, _, etc.)
    - Punctuation boundaries
    - Structural noise (bullets, numbering) across newlines

    REGEX SAFETY: All optional marker groups use atomic-group syntax (?>...)
    to prevent catastrophic backtracking. The previous non-atomic version
    `(?:\\*\\*|__|\\*|_)?` interleaved with `\\s+` produced exponential
    backtracking on long targets in long haystacks. Atomic groups commit
    on first match and never reconsider, making the regex linear.
    """
    target_text = _replace_smart_quotes(target_text)

    parts = []
    # Tokenize: Underscores, Whitespace, Quotes, AND Punctuation
    token_pattern = re.compile(r"(_+)|(\s+)|(['\"])|([.,;:])")

    # Pattern to allow optional markdown markers between tokens.
    # Atomic group prevents backtracking through marker permutations.
    md_noise = r"(?>\*\*|__|\*|_)*"

    # Pattern for Structural Noise (bullets, indentation, numbering)
    structural_noise = r"(?:\s*(?:[*+\->]|\d+\.)\s+|\s*\n\s*)"

    # START ANCHOR:
    # Allow optional list marker at the very start
    start_list_marker = r"(?:[ \t]*(?:[*+\->]|\d+\.)\s+)?"
    parts.append(start_list_marker)
    parts.append(md_noise)

    last_idx = 0
    for match in token_pattern.finditer(target_text):
        literal = target_text[last_idx : match.start()]
        if literal:
            parts.append(re.escape(literal))
            parts.append(md_noise)

        g_underscore, g_space, g_quote, g_punct = match.groups()

        if g_underscore:
            parts.append(r"_+")
        elif g_space:
            # If the whitespace contains a newline, allow structural noise
            if "\n" in g_space:
                parts.append(f"(?:{structural_noise}|\\s+)+")
            else:
                parts.append(r"\s+")
        elif g_quote:
            if g_quote == "'":
                parts.append(r"[\u2018\u2019']")
            else:
                parts.append(r"[\"\u201c\u201d]")
        elif g_punct:
            parts.append(re.escape(g_punct))

        parts.append(md_noise)
        last_idx = match.end()

    remaining = target_text[last_idx:]
    if remaining:
        parts.append(re.escape(remaining))

    return "".join(parts)


def _find_match_in_text(text: str, target: str) -> Tuple[int, int]:
    """
    Finds target in text using progressive matching strategies.
    Returns (start_idx, end_idx) or (-1, -1) if not found.
    """
    if not target:
        return -1, -1

    # 1. Exact match
    idx = text.find(target)
    if idx != -1:
        return _find_safe_boundaries(text, idx, idx + len(target))

    # 2. Smart quote normalization
    norm_text = _replace_smart_quotes(text)
    norm_target = _replace_smart_quotes(target)
    idx = norm_text.find(norm_target)
    if idx != -1:
        return _find_safe_boundaries(text, idx, idx + len(norm_target))

    # 3. Fuzzy regex match (handles markdown noise, list markers, etc.).
    # Atomic groups in _make_fuzzy_regex prevent catastrophic backtracking.
    try:
        pattern = _make_fuzzy_regex(target)
        match = re.search(pattern, text)
        if match:
            raw_start, raw_end = match.start(), match.end()
            refined_start, refined_end = _refine_match_boundaries(
                text, raw_start, raw_end
            )
            return _find_safe_boundaries(text, refined_start, refined_end)
    except re.error:
        pass

    return -1, -1


# Maximum number of match examples to include in an ambiguity error message.
# Capped to keep error payloads bounded — high-frequency tokens (like a brand
# name appearing 100+ times) would otherwise produce tens of KB of context
# snippets that consume LLM context budget without adding signal.
AMBIGUITY_EXAMPLES_CAP = 5

# Length of the surrounding-text window shown for each match example.
AMBIGUITY_CONTEXT_CHARS = 50


def format_ambiguity_error(
    edit_index: int,
    target_text: str,
    haystack: str,
    match_positions: list[tuple[int, int]],
) -> str:
    """
    Builds a uniformly-formatted ambiguity error message used by both the disk
    and Live Word edit pipelines.

    Args:
        edit_index: 1-based index of the failing edit, used in the message prefix.
        target_text: the search string the agent provided.
        haystack: the text the search was performed against.
        match_positions: list of (start, end) tuples for ALL matches found.
            The function shows up to AMBIGUITY_EXAMPLES_CAP examples and
            indicates how many additional matches are not shown.

    Returns:
        A multi-line error string suitable for inclusion in the
        BatchValidationError list (disk path) or skipped_details (Live Word path).

    Raises:
        ValueError: if match_positions has fewer than 2 entries (this helper is
            only meaningful for genuine ambiguity).
    """
    total = len(match_positions)
    if total < 2:
        raise ValueError(
            f"format_ambiguity_error requires at least 2 matches, got {total}"
        )

    shown = match_positions[:AMBIGUITY_EXAMPLES_CAP]
    remaining = total - len(shown)

    lines = [
        f"- Edit {edit_index} Failed: Ambiguous match. Target text appears "
        f"{total} times. First {len(shown)} occurrences:"
    ]

    for i, (start, end) in enumerate(shown, start=1):
        pre_start = max(0, start - AMBIGUITY_CONTEXT_CHARS)
        post_end = min(len(haystack), end + AMBIGUITY_CONTEXT_CHARS)

        pre_context = haystack[pre_start:start].replace("\n", " ")
        post_context = haystack[end:post_end].replace("\n", " ")
        match_text = haystack[start:end].replace("\n", " ")

        # Truncate displayed match itself if pathologically long.
        if len(match_text) > 50:
            match_text = match_text[:25] + "..." + match_text[-20:]

        prefix_marker = "..." if pre_start > 0 else ""
        suffix_marker = "..." if post_end < len(haystack) else ""

        lines.append(
            f'    {i}. "{prefix_marker}{pre_context}[{match_text}]{post_context}{suffix_marker}"'
        )

    if remaining > 0:
        lines.append(f"    ... and {remaining} more occurrence(s) not shown.")

    lines.append(
        "  Please provide more surrounding context in your target_text "
        "to uniquely identify the location."
    )

    return "\n".join(lines)


def _build_critic_markup(
    target_text: str,
    new_text: str,
    comment: Optional[str],
    edit_index: int,
    include_index: bool,
    highlight_only: bool,
) -> str:
    """
    Generates CriticMarkup string for a single edit.
    """
    parts = []

    # Strip balanced markers from target
    prefix_markup, clean_target, suffix_markup = _strip_balanced_markers(target_text)

    # If we stripped markers from target, try to strip the SAME markers from new_text
    clean_new = new_text
    if prefix_markup and new_text:
        # Check if new_text has the same outer markers
        if new_text.startswith(prefix_markup) and new_text.endswith(suffix_markup):
            inner_len = len(prefix_markup)
            clean_new = (
                new_text[inner_len:-inner_len]
                if len(new_text) > inner_len * 2
                else new_text
            )

    parts.append(prefix_markup)

    if highlight_only:
        parts.append(f"{{=={clean_target}==}}")
    else:
        has_target = bool(clean_target)
        has_new = bool(clean_new)

        if has_target and not has_new:
            parts.append(f"{{--{clean_target}--}}")
        elif not has_target and has_new:
            parts.append(f"{{++{clean_new}++}}")
        elif has_target and has_new:
            parts.append(f"{{--{clean_target}--}}{{++{clean_new}++}}")

    parts.append(suffix_markup)

    # Build metadata block
    meta_parts = []
    if comment:
        meta_parts.append(comment)
    if include_index:
        meta_parts.append(f"[Edit:{edit_index}]")

    if meta_parts:
        meta_content = " ".join(meta_parts)
        parts.append(f"{{>>{meta_content}<<}}")

    return "".join(parts)


def apply_edits_to_markdown(
    markdown_text: str,
    edits: List[ModifyText],
    include_index: bool = False,
    highlight_only: bool = False,
) -> str:
    """
    Applies edits to Markdown text and returns CriticMarkup-annotated output.
    """
    if not edits:
        return markdown_text

    # Step 1: Find match positions for each edit
    matched_edits: List[Tuple[int, int, str, ModifyText, int]] = []

    for idx, edit in enumerate(edits):
        target = edit.target_text or ""

        if not target:
            if highlight_only:
                logger.debug(
                    f"Skipping edit {idx}: no target_text in highlight_only mode"
                )
                continue
            else:
                logger.warning(
                    f"Skipping edit {idx}: pure insertion without target_text not supported in text mode"
                )
                continue

        start, end = _find_match_in_text(markdown_text, target)

        if start == -1:
            logger.warning(
                f"Skipping edit {idx}: target_text not found: '{target[:50]}...'"
            )
            continue

        actual_matched_text = markdown_text[start:end]
        matched_edits.append((start, end, actual_matched_text, edit, idx))

    # Step 2: Check for overlapping edits
    matched_edits_filtered: List[Tuple[int, int, str, ModifyText, int]] = []
    occupied_ranges: List[Tuple[int, int]] = []

    matched_edits.sort(key=lambda x: x[4])

    for start, end, actual_text, edit, orig_idx in matched_edits:
        overlaps = False
        for occ_start, occ_end in occupied_ranges:
            if start < occ_end and end > occ_start:
                overlaps = True
                logger.warning(
                    f"Skipping edit {orig_idx}: overlaps with previously matched edit"
                )
                break

        if not overlaps:
            matched_edits_filtered.append((start, end, actual_text, edit, orig_idx))
            occupied_ranges.append((start, end))

    # Step 3: Sort by position descending
    matched_edits_filtered.sort(key=lambda x: x[0], reverse=True)

    # Step 4: Apply edits
    result = markdown_text

    for start, end, actual_text, edit, orig_idx in matched_edits_filtered:
        new = edit.new_text or ""

        # Apply context trimming to isolate the actual change from the anchor
        prefix_len, suffix_len = trim_common_context(actual_text, new)

        # Extract the unmodified prefix and suffix
        unmodified_prefix = actual_text[:prefix_len] if prefix_len > 0 else ""
        unmodified_suffix = (
            actual_text[len(actual_text) - suffix_len :] if suffix_len > 0 else ""
        )

        # Isolate the actual target and new text to be marked up
        t_end = len(actual_text) - suffix_len
        n_end = len(new) - suffix_len
        isolated_target = actual_text[prefix_len:t_end]
        isolated_new = new[prefix_len:n_end]

        markup = _build_critic_markup(
            target_text=isolated_target,
            new_text=isolated_new,
            comment=edit.comment,
            edit_index=orig_idx,
            include_index=include_index,
            highlight_only=highlight_only,
        )

        # Recombine the unmodified anchors with the newly generated markup block
        full_replacement = unmodified_prefix + markup + unmodified_suffix
        result = result[:start] + full_replacement + result[end:]

    return result

File: node/packages/core/src/markup.ts
import { trim_common_context } from "./diff.js";
import { ModifyText } from "./models.js";
export const AMBIGUITY_EXAMPLES_CAP = 5;
export const AMBIGUITY_CONTEXT_CHARS = 50;
function _should_strip_markers(text: string, marker: string): boolean {
  if (!text.startsWith(marker) || !text.endsWith(marker)) return false;
  if (text.length < marker.length * 2) return false;

  const inner = text.substring(marker.length, text.length - marker.length);
  if (!inner) return false;

  if (inner.includes(marker)) return false;
  if (!/[a-zA-Z]/.test(inner)) return false;

  if (marker === "__" && /^\w+$/.test(inner)) return false;
  if (marker === "_") {
    if (inner.includes("_")) return false;
    if (/^[0-9_]+$/.test(inner)) return false;
  }

  return true;
}

function _strip_balanced_markers(text: string): [string, string, string] {
  let prefix_markup = "";
  let suffix_markup = "";
  let clean_text = text;

  const markers = ["**", "__", "_", "*"];

  for (const marker of markers) {
    if (_should_strip_markers(clean_text, marker)) {
      prefix_markup += marker;
      suffix_markup = marker + suffix_markup;
      clean_text = clean_text.substring(
        marker.length,
        clean_text.length - marker.length,
      );
      break;
    }
  }

  return [prefix_markup, clean_text, suffix_markup];
}

export function _replace_smart_quotes(text: string): string {
  return text
    .replace(/“/g, '"')
    .replace(/”/g, '"')
    .replace(/‘/g, "'")
    .replace(/’/g, "'");
}

function _find_safe_boundaries(
  text: string,
  start: number,
  end: number,
): [number, number] {
  let new_start = start;
  let new_end = end;

  const expand_if_unbalanced = (marker: string) => {
    const current_match = text.substring(new_start, new_end);
    const count = (
      current_match.match(new RegExp(marker.replace(/\*/g, "\\*"), "g")) || []
    ).length;

    if (count % 2 !== 0) {
      const suffix = text.substring(new_end);
      if (suffix.startsWith(marker)) {
        new_end += marker.length;
        return;
      }
      const prefix = text.substring(0, new_start);
      if (prefix.endsWith(marker)) {
        new_start -= marker.length;
        return;
      }
    }
  };

  for (let i = 0; i < 2; i++) {
    expand_if_unbalanced("**");
    expand_if_unbalanced("__");
    expand_if_unbalanced("_");
    expand_if_unbalanced("*");
  }

  return [new_start, new_end];
}

function _refine_match_boundaries(
  text: string,
  start: number,
  end: number,
): [number, number] {
  const markers = ["**", "__", "*", "_"];
  let current_text = text.substring(start, end);
  let best_start = start;
  let best_end = end;

  const countMarker = (str: string, mk: string) =>
    (str.match(new RegExp(mk.replace(/\*/g, "\\*"), "g")) || []).length;

  for (const marker of markers) {
    if (current_text.startsWith(marker)) {
      const current_score = countMarker(current_text, marker) % 2;
      const trimmed_text = current_text.substring(marker.length);
      const trimmed_score = countMarker(trimmed_text, marker) % 2;

      if (current_score === 1 && trimmed_score === 0) {
        best_start += marker.length;
        current_text = trimmed_text;
      }
    }
  }

  for (const marker of markers) {
    if (current_text.endsWith(marker)) {
      const current_score = countMarker(current_text, marker) % 2;
      const trimmed_text = current_text.substring(
        0,
        current_text.length - marker.length,
      );
      const trimmed_score = countMarker(trimmed_text, marker) % 2;

      if (current_score === 1 && trimmed_score === 0) {
        best_end -= marker.length;
        current_text = trimmed_text;
      }
    }
  }

  return [best_start, best_end];
}

export function _make_fuzzy_regex(target_text: string): string {
  target_text = _replace_smart_quotes(target_text);

  const parts: string[] = [];
  const token_pattern = /(_+)|(\s+)|(['"])|([.,;:\/])/g;

  // Note: JS does not support atomic groups (?>...).
  // However, because we only match markdown characters * and _,
  // we can use a character class `[*_]*` which is mathematically equivalent
  // to `(?:\*\*|__|\*|_)*` but fundamentally immune to catastrophic backtracking!
  const md_noise = "[*_]*";
  const structural_noise = "(?:\\s*(?:[*+\\->]|\\d+\\.)\\s+|\\s*\\n\\s*)";

  const start_list_marker = "(?:[ \\t]*(?:[*+\\->]|\\d+\\.)\\s+)?";
  parts.push(start_list_marker);
  parts.push(md_noise);

  let last_idx = 0;
  let match;

  const escapeRegExp = (str: string) =>
    str.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");

  while ((match = token_pattern.exec(target_text)) !== null) {
    const literal = target_text.substring(last_idx, match.index);
    if (literal) {
      parts.push(escapeRegExp(literal));
      parts.push(md_noise);
    }

    const g_underscore = match[1];
    const g_space = match[2];
    const g_quote = match[3];
    const g_punct = match[4];

    if (g_underscore) {
      parts.push("_+");
    } else if (g_space) {
      if (g_space.includes("\n")) {
        parts.push(`(?:${structural_noise}|\\s+)+`);
      } else {
        parts.push("\\s+");
      }
    } else if (g_quote) {
      if (g_quote === "'") parts.push("[\u2018\u2019']");
      else parts.push('["\u201c\u201d]');
    } else if (g_punct) {
      parts.push(escapeRegExp(g_punct));
    }

    parts.push(md_noise);
    last_idx = token_pattern.lastIndex;
  }

  const remaining = target_text.substring(last_idx);
  if (remaining) parts.push(escapeRegExp(remaining));

  return parts.join("");
}

export function _find_match_in_text(
  text: string,
  target: string,
): [number, number] {
  if (!target) return [-1, -1];

  let idx = text.indexOf(target);
  if (idx !== -1) return _find_safe_boundaries(text, idx, idx + target.length);

  const norm_text = _replace_smart_quotes(text);
  const norm_target = _replace_smart_quotes(target);
  idx = norm_text.indexOf(norm_target);
  if (idx !== -1)
    return _find_safe_boundaries(text, idx, idx + norm_target.length);

  try {
    const pattern = new RegExp(_make_fuzzy_regex(target));
    const match = pattern.exec(text);
    if (match) {
      const raw_start = match.index;
      const raw_end = match.index + match[0].length;
      const [refined_start, refined_end] = _refine_match_boundaries(
        text,
        raw_start,
        raw_end,
      );
      return _find_safe_boundaries(text, refined_start, refined_end);
    }
  } catch (e) {
    // Ignore regex compilation errors from edge cases
  }

  return [-1, -1];
}

export function _build_critic_markup(
  target_text: string,
  new_text: string,
  comment: string | null | undefined,
  edit_index: number,
  include_index: boolean,
  highlight_only: boolean,
): string {
  const parts: string[] = [];

  let [prefix_markup, clean_target, suffix_markup] =
    _strip_balanced_markers(target_text);

  let clean_new = new_text;
  if (prefix_markup && new_text) {
    if (
      new_text.startsWith(prefix_markup) &&
      new_text.endsWith(suffix_markup)
    ) {
      const inner_len = prefix_markup.length;
      clean_new =
        new_text.length > inner_len * 2
          ? new_text.substring(inner_len, new_text.length - inner_len)
          : new_text;
    }
  }

  parts.push(prefix_markup);

  if (highlight_only) {
    parts.push(`{==${clean_target}==}`);
  } else {
    const has_target = Boolean(clean_target);
    const has_new = Boolean(clean_new);

    if (has_target && !has_new) parts.push(`{--${clean_target}--}`);
    else if (!has_target && has_new) parts.push(`{++${clean_new}++}`);
    else if (has_target && has_new)
      parts.push(`{--${clean_target}--}{++${clean_new}++}`);
  }

  parts.push(suffix_markup);

  const meta_parts: string[] = [];
  if (comment) meta_parts.push(comment);
  if (include_index) meta_parts.push(`[Edit:${edit_index}]`);

  if (meta_parts.length > 0) {
    parts.push(`{>>${meta_parts.join(" ")}<<}`);
  }

  return parts.join("");
}

export function apply_edits_to_markdown(
  markdown_text: string,
  edits: ModifyText[],
  include_index = false,
  highlight_only = false,
): string {
  if (!edits || edits.length === 0) return markdown_text;

  const matched_edits: [number, number, string, ModifyText, number][] = [];

  for (let idx = 0; idx < edits.length; idx++) {
    const edit = edits[idx];
    const target = edit.target_text || "";

    if (!target) {
      continue;
    }

    const [start, end] = _find_match_in_text(markdown_text, target);
    if (start === -1) continue;

    const actual_matched_text = markdown_text.substring(start, end);
    matched_edits.push([start, end, actual_matched_text, edit, idx]);
  }

  const matched_edits_filtered: [number, number, string, ModifyText, number][] =
    [];
  const occupied_ranges: [number, number][] = [];

  matched_edits.sort((a, b) => a[4] - b[4]);

  for (const [start, end, actual_text, edit, orig_idx] of matched_edits) {
    let overlaps = false;
    for (const [occ_start, occ_end] of occupied_ranges) {
      if (start < occ_end && end > occ_start) {
        overlaps = true;
        break;
      }
    }

    if (!overlaps) {
      matched_edits_filtered.push([start, end, actual_text, edit, orig_idx]);
      occupied_ranges.push([start, end]);
    }
  }

  matched_edits_filtered.sort((a, b) => b[0] - a[0]);

  let result = markdown_text;

  for (const [
    start,
    end,
    actual_text,
    edit,
    orig_idx,
  ] of matched_edits_filtered) {
    const new_txt = edit.new_text || "";
    const [prefix_len, suffix_len] = trim_common_context(actual_text, new_txt);

    const unmodified_prefix =
      prefix_len > 0 ? actual_text.substring(0, prefix_len) : "";
    const unmodified_suffix =
      suffix_len > 0
        ? actual_text.substring(actual_text.length - suffix_len)
        : "";

    const t_end = actual_text.length - suffix_len;
    const n_end = new_txt.length - suffix_len;

    const isolated_target = actual_text.substring(prefix_len, t_end);
    const isolated_new = new_txt.substring(prefix_len, n_end);

    const markup = _build_critic_markup(
      isolated_target,
      isolated_new,
      edit.comment,
      orig_idx,
      include_index,
      highlight_only,
    );

    const full_replacement = unmodified_prefix + markup + unmodified_suffix;
    result =
      result.substring(0, start) + full_replacement + result.substring(end);
  }

  return result;
}
export function format_ambiguity_error(
  edit_index: number,
  target_text: string,
  haystack: string,
  match_positions: [number, number][],
): string {
  const total = match_positions.length;
  if (total < 2) {
    throw new Error(
      `format_ambiguity_error requires at least 2 matches, got ${total}`,
    );
  }

  const shown = match_positions.slice(0, AMBIGUITY_EXAMPLES_CAP);
  const remaining = total - shown.length;

  const lines: string[] = [
    `- Edit ${edit_index} Failed: Ambiguous match. Target text appears ${total} times. First ${shown.length} occurrences:`,
  ];

  for (let i = 0; i < shown.length; i++) {
    const [start, end] = shown[i];
    const pre_start = Math.max(0, start - AMBIGUITY_CONTEXT_CHARS);
    const post_end = Math.min(haystack.length, end + AMBIGUITY_CONTEXT_CHARS);

    const pre_context = haystack
      .substring(pre_start, start)
      .replace(/\n/g, " ");
    const post_context = haystack.substring(end, post_end).replace(/\n/g, " ");
    let match_text = haystack.substring(start, end).replace(/\n/g, " ");

    if (match_text.length > 50) {
      match_text =
        match_text.substring(0, 25) +
        "..." +
        match_text.substring(match_text.length - 20);
    }

    const prefix_marker = pre_start > 0 ? "..." : "";
    const suffix_marker = post_end < haystack.length ? "..." : "";

    lines.push(
      `    ${i + 1}. "${prefix_marker}${pre_context}[${match_text}]${post_context}${suffix_marker}"`,
    );
  }

  if (remaining > 0) {
    lines.push(`    ... and ${remaining} more occurrence(s) not shown.`);
  }

  // Tell the agent EXACTLY how to re-call. Without this, agents loop forever
  // refining target_text/regex because they never learn that match_mode is the
  // built-in escape hatch for genuine ambiguity.
  lines.push("  To resolve, re-send this edit using ONE of these strategies:");
  lines.push(
    `    1. Set "match_mode": "all" to modify ALL ${total} occurrences (same target_text).`,
  );
  lines.push(
    '    2. Set "match_mode": "first" to modify only the FIRST occurrence (same target_text).',
  );
  lines.push(
    '    3. Please provide more surrounding context in your target_text to uniquely ' +
      'identify a single location (keep the default "match_mode": "strict").',
  );

  return lines.join("\n");
}
